### PR TITLE
fix image size calculation in OBB and Pose26 loss for multi-scale training

### LINKED ISSUE #23232: YOLO26n-obb训练无法收敛
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

同一个obb数据集在yolo11上3块gpu训练结果正常和良好

但是用最新的YOLO26n-obb训练，就一直无法收敛（单个GPU和多个GPU都测试过）

不知道是不是跟图片很大有关系？imgsz=4576

训练代码：

from ultralytics import YOLO
model = YOLO('yolo26n-obb.pt') 
results = model.train(
    data="/home/roboscope/ai_model_train/microbe/datasets/20251125195516_4576_20260107102001/data.yaml", 
    epochs=500, 
    imgsz=4576,
    device=[1,2,3],
    project="/home/roboscope/ai_model_train/microbe/results/scope-s1_all_4576",
    name="yolo26n_4576",
    patience=100,
    batch=4,
    save_period=20,
    cache="disk",
    workers=2,
    amp=True,
    multi_scale=False,
    exist_ok=False,
    pretrained=True,
    optimizer="AdamW",
    max_det=99999,
    )

输出是：
YOLO26n-obb summary: 291 layers, 2,649,474 parameters, 2,649,474 gradients, 6.3 GFLOPs

Transferred 780/792 items from pretrained weights
DDP: debug command /home/roboscope/miniconda3/envs/yolov11/bin/python -m torch.distributed.run --nproc_per_node 3 --master_port 50167 /home/roboscope/.config/Ultralytics/DDP/_temp_hcgwurs7138840300637296.py
Ultralytics 8.4.2 🚀 Python-3.9.25 torch-2.8.0+cu128 CUDA:1 (NVIDIA A800 80GB PCIe, 81154MiB)
                                                    CUDA:2 (NVIDIA A800 80GB PCIe, 81154MiB)
                                                    CUDA:3 (NVIDIA A800 80GB PCIe, 81154MiB)
Overriding model.yaml nc=15 with nc=6
Transferred 780/792 items from pretrained weights
AMP: running Automatic Mixed Precision (AMP) checks...
AMP: checks passed ✅
train: Fast image access ✅ (ping: 0.0±0.0 ms, read: 693.8±310.5 MB/s, size: 8365.1 KB)
train: Scanning /home/roboscope/ai_model_train/microbe/datasets/20251125195516_4576_20260107102001/train/labels.cache... 1351 images, 305 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1351/1351 182.8Mit/s 0.0s
train: Caching images (79.0GB Disk): 100% ━━━━━━━━━━━━ 1351/1351 12.1Kit/s 0.1s
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1081.3±585.1 MB/s, size: 6282.5 KB)
val: Scanning /home/roboscope/ai_model_train/microbe/datasets/20251125195516_4576_20260107102001/val/labels.cache... 600 images, 141 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 600/600 22.1Mit/s 0.0s
val: Caching images (35.1GB Disk): 100% ━━━━━━━━━━━━ 600/600 15.1Kit/s 0.0s
Plotting labels to /home/roboscope/ai_model_train/microbe/results/scope-s1_all_4576/yolo26n_457613/labels.jpg... 
optimizer: AdamW(lr=0.01, momentum=0.937) with parameter groups 126 weight(decay=0.0), 144 weight(decay=0.0005), 144 bias(decay=0.0)
Image sizes 4576 train, 4576 val
Using 6 dataloader workers
Logging results to /home/roboscope/ai_model_train/microbe/results/scope-s1_all_4576/yolo26n_457613
Starting training for 500 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      1/500      29.1G      1.757      6.564   0.003644     0.2816          4       4576: 100% ━━━━━━━━━━━━ 451/451 2.1it/s 3:37
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 10.2it/s 19.7s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      2/500      29.1G       1.49      2.963   0.002829     0.2166          3       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:19
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.6it/s 17.3s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      3/500      29.1G      1.436      2.777   0.002788      0.217          1       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:17
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.5it/s 17.3s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      4/500      29.1G      1.399      2.614   0.002837     0.2126          6       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:16
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.6it/s 17.2s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      5/500      29.2G      1.353      2.648   0.002829     0.2024         28       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:16
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.5it/s 17.4s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      6/500      29.2G      1.386      2.521   0.002929     0.1988         13       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:16
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.8it/s 16.9s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      7/500      36.3G       1.32      2.413     0.0025     0.1875         27       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:15
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.8it/s 16.9s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      8/500      36.3G      1.321       2.44   0.002673     0.1873          2       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:15
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.7it/s 17.1s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
      9/500      36.3G        nan        nan        nan        nan         15       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:17
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.0it/s 18.2s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
     10/500      36.3G        nan        nan        nan        nan         11       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:18
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.1it/s 18.0s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
     11/500      36.3G      1.368      2.698   0.002855     0.1907         12       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:17
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.2it/s 17.9s
                   all        600      22564          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss angle_loss  Instances       Size
     12/500      36.3G      1.307      2.624   0.002525     0.1963         14       4576: 100% ━━━━━━━━━━━━ 451/451 2.3it/s 3:17
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 200/200 11.2it/s 17.9s

### Environment

Ultralytics 8.4.2 🚀 Python-3.9.25 torch-2.8.0+cu128 
CUDA:1 (NVIDIA A800 80GB PCIe, 81154MiB)
CUDA:2 (NVIDIA A800 80GB PCIe, 81154MiB)
CUDA:3 (NVIDIA A800 80GB PCIe, 81154MiB)

### Minimal Reproducible Example

from ultralytics import YOLO
model = YOLO('yolo26n-obb.pt') 
results = model.train(
    data="/home/roboscope/ai_model_train/microbe/datasets/20251125195516_4576_20260107102001/data.yaml", 
    epochs=500, 
    imgsz=4576,
    device=[1,2,3],
    project="/home/roboscope/ai_model_train/microbe/results/scope-s1_all_4576",
    name="yolo26n_4576",
    patience=100,
    batch=4,
    save_period=20,
    cache="disk",
    workers=2,
    amp=True,
    multi_scale=False,
    exist_ok=False,
    pretrained=True,
    optimizer="AdamW",
    max_det=99999,
    )

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by lmycross**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes how image size is computed inside YOLO losses by deriving it from model feature maps instead of relying on batch metadata 🛠️📐

### 📊 Key Changes
- Updated loss code to compute `imgsz` from `preds["feats"][0].shape[2:] * self.stride[0]` rather than `batch["resized_shape"][0]` 🔄
- Applied this change in multiple loss paths, including:
  - Keypoint/pose-related loss (uses `pred_kpts`) 🧍‍♂️
  - Angle/rotation-related loss (uses `pred_angle`, e.g., oriented boxes) 📦🧭

### 🎯 Purpose & Impact
- Improves robustness: loss no longer depends on `batch["resized_shape"]` being present or perfectly accurate ✅
- Ensures the loss uses the *actual effective network input size* inferred from feature maps + stride, which is especially helpful with varying preprocessing pipelines or custom dataloaders 🧩
- Reduces risk of subtle scaling mismatches in pose and rotated-object training, leading to more consistent training behavior 🔧📈

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lmycross, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message, and an Ultralytics engineer will assist you shortly 😊 To ensure a seamless integration of your work, please review the following checklist:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

PR looks clean. The new image-size computation aligns with feature-map resolution and doesn’t introduce obvious errors.

**Review by Laughing-q — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Merged — awesome work @lmycross, and thank you @Laughing-q for the contributions! 🙌

“**If I have seen further it is by standing on the shoulders of giants.**” — *Isaac Newton*  
This PR is a perfect example of that mindset: by deriving `imgsz` directly from the model’s feature maps + stride, you’ve made the YOLO loss logic more grounded in what the network actually sees—less dependent on fragile batch metadata, and more reliable across custom dataloaders and preprocessing pipelines. The improvements for pose/keypoints and oriented boxes especially help prevent subtle scaling mismatches and keep training behavior consistent.

Appreciate the careful engineering and the impact on robustness—thanks for helping make Ultralytics and **YOLO26** stronger for everyone.
